pkg/v2/crud/traverse.go

```diff
@@ -2,21 +2,52 @@ package crud
 
 import (
 	"fmt"
+
 	"github.com/imulab/go-scim/pkg/v2/annotation"
 	"github.com/imulab/go-scim/pkg/v2/crud/expr"
 	"github.com/imulab/go-scim/pkg/v2/prop"
 	"github.com/imulab/go-scim/pkg/v2/spec"
 )
 
-func defaultTraverse(property prop.Property, query *expr.Expression, callback func(nav prop.Navigator) error) error {
+type traverseCb func(nav prop.Navigator, value interface{}) error
+
+func defaultTraverse(property prop.Property, query *expr.Expression, callback traverseCb) error {
+	tr := traverser{
+		nav:             prop.Navigate(property),
+		callback:        callback,
+		elementStrategy: selectAllStrategy,
+	}
+	return tr.traverse(query)
+}
+
+// A single 'Eq' filter can be used to add a new attribute.
+// This traverse calls the callback with the modified value using such filter.
+// The operation like:
+//
+//	{
+//		"op": "add",
+//		"path": "emails[type eq \"work\"].value",
+//		"value": "foo@bar.com"
+//	}
+//
+// Adds a new property:
+//
+//	"emails": [
+//		{
+//			"type": "work",
+//			"value": "foo@bar.com"
+//		}
+//	]
+func addByEqFilterTraverse(value interface{}, property prop.Property, query *expr.Expression, callback traverseCb) error {
 	return traverser{
```

Comment:
thanks for this generalized change, however i think what i was after was a little different. instead of changing the signature of traverseCb everywhere, can you keep it the same but define a decorator for it here:

```
func addByEqFilterTraverse(value interface{}, property prop.Property, query *expr.Expression, callback traverseCb) error {
  modifierCb := func(nav prop.Navigator) error {
    nav.Add(value)
    traverseCb(nav) // if needed?
  }
  return traverser{
		nav:             prop.Navigate(property),
		callback:        modifierCb,
		elementStrategy: selectAllStrategy,
	}.traverse(query)
```

looks like we'd need a similar one for the update method below

Reply:
@vocoded I've added the decorators to keep the external callback signature not changed.
In order to compose a new value using data parsed fro query - I need:
- The `value` passed into `addByEqFilterTraverse`
- The modified `query` - traversed until a single Eq filter found
- The original navigator - to verify the values parsed from the `query`
The way you've proposed misses the modified `query`; that's why I've added one more strategy pattern to define when the traverse is done.
-

Reply:
ugh, ok. i thought that the decorator would make the diff considerably smaller by avoiding any changes to the traverser struct. in fact it is more involved and you still had to modify the traverser.

unfortunately i don't have sufficient context on this library or how it works to make any better suggestions. it seems surprising that we are adding a mutation operation (add) to a read operation (traverse) - conceptually i would want to separate them - traverse to the node that need to be updated, return it, then modify from the caller. so instead of above:

```
return addByEqFilterTraverse(value, resource.RootProperty(), skipMainSchemaNamespace(resource, head), func(nav prop.Navigator) error {
		return nav.Error()
	})
```

we could do:

```
nav := traverseByEqFilter(resource.RootProperty(), skipMainSchemaNamespace(resource, head))
nav.Add(value)
```

if this is not possible i'd revert to your previous commit which was a little less complicated. sorry for the back and forth, just trying to get to the simplest solution that will work, which will be easier to maintain

Reply:
It's possible; but it differs from the other traverse methods there.
I'd just use a different type of callback for addByEqFilterTraverse to return the updated `value` along with the properly focused `nav`.
Let me update the PR instead of describing this here :)

Reply:
@vocoded I've updated the PR according to the previous comment ^
take a look please

Reply:
it feels like we are still mixing concerns by combining traversal and value creation. ideally i'd expect there to be a sequence like:

```
// 1. Get value based on eqFilter
// 2. Create callback that will apply new value
// 3. Pass callback to a traversal that will stop on the right query expression (do we need a separate "traversalStrategy"?)
```
i think the library authors would probably reject this patch that seems to be doing an update on an Add operation, so i'm curious if they missed a case in the SCIM specification or this is special behavior we've observed we needed.

i don't want to burn more of your time on this because i know it's blocking a needed fix. assuming it works (are there tests we could write to verify?) then we can merge and come back to an improved version later.

Reply:
The thing is that traversing to the eq filter gives us both properly focused navigator and the validated query which could be used for value generation. I think I can traverse two times to satisfy the sequence you have suggested, not sure how elegant it’d be though

Reply:
@vocoded 
> if they missed a case in the SCIM specification or this is special behavior we've observed we needed.

Looks like the authors missed this from the [specification](https://datatracker.ietf.org/doc/html/rfc7644#section-3.5.2.1):

> The value MAY be a quoted value, or it may be
   a JSON object containing the sub-attributes of the complex attribute
   specified in the operation's "path".

>  If the target location does not exist, the attribute and value are
      added.

> If the target location specifies an attribute that does not exist
      (has no value), the attribute is added with the new value.